### PR TITLE
"Unread messages" filter

### LINKED ISSUE #178: manage unseen threads filter

### PR DESCRIPTION
**Pull request by FabianDevel**
Depends on #179 
Closes #178 

Add filter to enable the unread button in ThreadsList.
`UpdatedAt` field in ThreadsList header is now working.

### CODE REVIEW

**Review by JorisBodin — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt [hunk 1]
```diff
@@ -151,20 +148,31 @@ class ThreadListFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {
     }
 
     private fun updateUpdatedAt() = with(binding) {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt:150]:
```suggestion
    private fun updateUpdatedAt() {
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt [hunk 2]
```diff
@@ -151,20 +148,31 @@ class ThreadListFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {
...
-        val ago = if (Date().time - lastUpdatedAt.time < DateUtils.MINUTE_IN_MILLIS) {
-            getString(R.string.threadListHeaderLastUpdateNow)
+        val lastUpdatedAt = viewModel.currentFolder.value?.lastUpdatedAt?.toDate()
+        val ago = when {
+            lastUpdatedAt == null -> ""
+            Date().time - lastUpdatedAt.time < DateUtils.MINUTE_IN_MILLIS -> getString(R.string.threadListHeaderLastUpdateNow)
+            else -> DateUtils.getRelativeTimeSpanString(lastUpdatedAt.time).toString().replaceFirstChar { it.lowercaseChar() }
+        }
+
+        updatedAt.text = if (ago.isEmpty()) {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt:158]:
```suggestion
        val folder = viewModel.currentFolder.value ?: return
        val lastUpdatedAt = folder.lastUpdatedAt?.toDate()
        val ago = when {
            lastUpdatedAt == null -> ""
            Date().time - lastUpdatedAt.time < DateUtils.MINUTE_IN_MILLIS -> getString(R.string.threadListHeaderLastUpdateNow)
            else -> DateUtils.getRelativeTimeSpanString(lastUpdatedAt.time).toString().replaceFirstChar { it.lowercaseChar() }
        }

        binding.updatedAt.text = if (ago.isEmpty()) {
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt [hunk 3]
```diff
@@ -151,20 +148,31 @@ class ThreadListFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {
     }
 
     private fun updateUpdatedAt() = with(binding) {
-        // TODO : Replace lastUpdatedAt.time with currentFolder.lastUpdatedAt ?
-        val ago = if (Date().time - lastUpdatedAt.time < DateUtils.MINUTE_IN_MILLIS) {
-            getString(R.string.threadListHeaderLastUpdateNow)
+        val lastUpdatedAt = viewModel.currentFolder.value?.lastUpdatedAt?.toDate()
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt:151]:
```suggestion
        val folder = viewModel.currentFolder.value ?: return
        val lastUpdatedAt = folder.lastUpdatedAt?.toDate()
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt (hunk 2 above)
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt:158]:
```suggestion
        binding.updatedAt.text = if (ago.isEmpty()) {
```

File: app/src/main/java/com/infomaniak/mail/data/api/ApiRoutes.kt [hunk 4]
```diff
@@ -18,6 +18,7 @@
 package com.infomaniak.mail.data.api
 
 import com.infomaniak.mail.BuildConfig.MAIL_API
+import com.infomaniak.mail.data.models.thread.Thread
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/data/api/ApiRoutes.kt:21]:
```suggestion
import com.infomaniak.mail.data.models.thread.Thread.ThreadFilter
```

File: app/src/main/java/com/infomaniak/mail/data/api/ApiRoutes.kt [hunk 5]
```diff
@@ -47,8 +48,17 @@ object ApiRoutes {
...
 
-    fun threads(uuid: String, folderId: String, offset: Int, filter: String?): String {
-        return "${folder(uuid, folderId)}/message?thread=on&offset=$offset&filter=$filter"
+    fun threads(uuid: String, folderId: String, offset: Int, filter: Thread.ThreadFilter, searchText: String? = null): String {
+        val urlSearch = searchText?.let { "&scontains=$it" } ?: ""
+        val urlAttachment = if (filter == Thread.ThreadFilter.ATTACHMENTS) "&sattachments=yes" else ""
+        val urlFilter = when (filter) {
+            Thread.ThreadFilter.SEEN,
+            Thread.ThreadFilter.STARRED,
+            Thread.ThreadFilter.UNSEEN -> "&filters=${filter.name.lowercase()}"
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/data/api/ApiRoutes.kt:57]:
```suggestion
    fun threads(uuid: String, folderId: String, offset: Int, filter: ThreadFilter, searchText: String? = null): String {
        val urlSearch = searchText?.let { "&scontains=$it" } ?: ""
        val urlAttachment = if (filter == ThreadFilter.ATTACHMENTS) "&sattachments=yes" else ""
        val urlFilter = when (filter) {
            ThreadFilter.SEEN,
            ThreadFilter.STARRED,
            ThreadFilter.UNSEEN -> "&filters=${filter.name.lowercase()}"
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt (hunk 1 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt:150]: Update binding #1/2

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt (hunk 2 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt:158]: Update binding 2/2

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt (hunk 3 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListFragment.kt:151]: Fix `lastUpdatedAt` blinking 1/1

File: app/src/main/java/com/infomaniak/mail/data/api/ApiRoutes.kt (hunk 4 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/api/ApiRoutes.kt:21]: Update `ThreadFilter` import 1/2

File: app/src/main/java/com/infomaniak/mail/data/api/ApiRoutes.kt (hunk 5 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/api/ApiRoutes.kt:57]: Update `ThreadFilter` import 2/2

File: app/src/main/java/com/infomaniak/mail/data/cache/MailboxContentController.kt [hunk 6]
```diff
@@ -58,9 +59,10 @@ object MailboxContentController {
 
     fun upsertFolder(folder: Folder): Folder = MailRealm.mailboxContent.writeBlocking { copyToRealm(folder, UpdatePolicy.ALL) }
 
-    // fun updateFolder(id: String, onUpdate: (folder: Folder) -> Unit) {
-    //     MailRealm.mailboxContent.writeBlocking { getLatestFolder(id)?.let(onUpdate) }
-    // }
+    fun updateFolder(id: String, onUpdate: (folder: Folder) -> Unit): Folder? {
+        MailRealm.mailboxContent.writeBlocking { getLatestFolder(id)?.let(onUpdate) }
```
[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/cache/MailboxContentController.kt:63]: Simplify `updateFolder` 1/2

File: app/src/main/java/com/infomaniak/mail/data/MailData.kt [hunk 7]
```diff
@@ -200,6 +204,14 @@ object MailData {
         }
     }
 
+    fun updateCurrentFolderCounts(folder: Folder, threadResult: ThreadsResult) {
+        mutableCurrentFolderFlow.value = MailboxContentController.updateFolder(folder.id) {
+            it.unreadCount = threadResult.folderUnseenMessage
+            it.totalCount = threadResult.totalMessagesCount
+            it.lastUpdatedAt = Date().toRealmInstant()
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/data/MailData.kt:211]:
```suggestion
            it.apply {
                unreadCount = threadResult.folderUnseenMessage
                totalCount = threadResult.totalMessagesCount
                lastUpdatedAt = Date().toRealmInstant()
            }
```

Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/MailData.kt:211]: Simplify `updateFolder` 2/2

File: app/src/main/java/com/infomaniak/mail/data/models/thread/Thread.kt [hunk 8]
```diff
@@ -99,12 +101,12 @@ class Thread : RealmObject {
...
-        UNSEEN("Unseen"),
-        STARRED("Starred"),
-        UNSTARRED("Unstarred"),
+    enum class ThreadFilter(@IdRes val filterNameRes: Int) {
+        ALL(R.string.searchAllMessages),
+        SEEN(R.string.searchFilterRead),
+        UNSEEN(R.string.searchFilterUnread),
+        STARRED(R.string.favoritesFolder),
+        ATTACHMENTS(R.string.searchFilterAttachment),
+        FOLDER(R.string.searchFilterFolder)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/data/models/thread/Thread.kt:110]:
```suggestion
        FOLDER(R.string.searchFilterFolder),
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListViewModel.kt [hunk 9]
```diff
@@ -54,17 +75,21 @@ class ThreadListViewModel : ViewModel() {
...
 
     fun refreshThreads() {
+        currentOffset = OFFSET_FIRST_PAGE
+        isDownloadingChanges = true
         MailData.refreshThreads(
-            folder = MailData.currentFolderFlow.value ?: return,
-            mailbox = MailData.currentMailboxFlow.value ?: return,
+            MailData.currentFolderFlow.value ?: return,
+            MailData.currentMailboxFlow.value ?: return,
+            filter
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadListViewModel.kt:92]:
```suggestion
            folder = MailData.currentFolderFlow.value ?: return,
            mailbox = MailData.currentMailboxFlow.value ?: return,
            filter = filter,
```

**Review by KevinBoulongne — APPROVED** [review]

**Pull request merged**